src/markov/ddp.jl

```diff
@@ -0,0 +1,468 @@
+function evaluate_policy(ddp::DiscreteDP, sigma::Vector{Int})
+    R_sigma, Q_sigma = RQ_sigma(ddp, sigma)
+    b = R_sigma
+    A = I - ddp.beta * Q_sigma
+    return A \ b
+end
+
+"""
+Solve the dynamic programming problem.
+
+##### Parameters
+
+- `ddp::DisreteDP` : Object that contains the Model Parameters
+- `method::Type{T<Algo}(VFI)`: Type name specifying solution method. Acceptable
+arguments are `VFI` for value function iteration or `PFI` for policy function
+iteration
+- `;max_iter::Int(250)` : Maximum number of iterations
+- `;epsilon::Float64(1e-3)` : Value for epsilon-optimality. Only used if
+`method` is `VFI`
+- `;k::Int(20)` : Number of iterations for partial policy evaluation in modified
+policy iteration (irrelevant for other methods).
+
+##### Returns
+
+- `ddpr::DPSolveResult{Algo}` : Optimization result represented as a
+DPSolveResult. See `DPSolveResult` for details.
+"""
+function solve{Algo<:DDPAlgorithm}(ddp::DiscreteDP, method::Type{Algo}=VFI;
+                                   max_iter::Int=250, epsilon::Float64=1e-3,
+                                   k::Int=20)
+    ddpr = DPSolveResult{Algo}(ddp)
+    _solve!(ddp, ddpr, max_iter, epsilon, k)
+    ddpr.mc = MarkovChain(ddp, ddpr)
+    ddpr
+end
+
+function solve{Algo<:DDPAlgorithm}(ddp::DiscreteDP, v_init::Vector,
+                                   method::Type{Algo}=VFI; max_iter::Int=250,
+                                   epsilon::Float64=1e-3, k::Int=20)
+    ddpr = DPSolveResult{Algo}(ddp, v_init)
+    _solve!(ddp, ddpr, max_iter, epsilon, k)
+    ddpr.mc = MarkovChain(ddp, ddpr)
+    ddpr
+end
+
+Base.ind2sub(ddp::DiscreteDP, ddpr::DPSolveResult) =
+    map(x->ind2sub(size(ddp.R), x)[2], ddpr.sigma)
+
+"""
+Impliments Value Iteration
+NOTE: See `solve` for further details
+"""
+function _solve!(ddp::DiscreteDP, ddpr::DPSolveResult{VFI}, max_iter::Int64,
+               epsilon::Float64, k::Int)
+    if ddp.beta == 0.0
+        tol = Inf
+    else
+        tol = epsilon * (1-ddp.beta) / (2*ddp.beta)
+    end
+
+    for i in 1:max_iter
+        # updates Tv in place
+        bellman_operator!(ddp, ddpr)
+
+        # compute error and update the v inside ddpr
+        err = maxabs(ddpr.Tv .- ddpr.v)
+        copy!(ddpr.v, ddpr.Tv)
+        ddpr.num_iter += 1
+
+        if err < tol
+            break
+        end
+    end
+
+    ddpr
+end
+
+"""
+Policy Function Iteration
+
+NOTE: The epsilon is ignored in this method. It is only here so dispatch can
+      go from `solve(::DiscreteDP, ::Type{Algo})` to any of the algorithms.
+      See `solve` for further details
+"""
+function _solve!(ddp::DiscreteDP, ddpr::DPSolveResult{PFI}, max_iter::Int,
+                epsilon::Float64, k::Int)
+    old_sigma = copy(ddpr.sigma)
+
+    for i in 1:max_iter
+       ddpr.v = evaluate_policy(ddp, ddpr)
+       compute_greedy!(ddp, ddpr)
+
+       if all(old_sigma .== ddpr.sigma)
+           break
+       end
+       copy!(old_sigma, ddpr.sigma)
+       ddpr.num_iter += 1
```

Comment:
This should come before the `if` block? (Or should initialize `ddpr` with `num_iter=1`?)